hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/ViewFileSystem.java

```diff
@@ -1211,13 +1211,29 @@ public FileStatus getFileStatus(Path f) throws IOException {
         INode<FileSystem> inode = iEntry.getValue();
         if (inode.isLink()) {
           INodeLink<FileSystem> link = (INodeLink<FileSystem>) inode;
-
-          result[i++] = new FileStatus(0, false, 0, 0,
-            creationTime, creationTime, PERMISSION_555,
-            ugi.getShortUserName(), ugi.getPrimaryGroupName(),
-            link.getTargetLink(),
-            new Path(inode.fullPath).makeQualified(
-                myUri, null));
+          // For MERGE or NFLY links, the first target link is considered
+          // for fetching the FileStatus with an assumption that the permission
+          // and the owner will be the same for all the target directories.
+          Path linkedPath = new Path(link.targetDirLinkList[0].toString());
+          ChRootedFileSystem linkedFs = (ChRootedFileSystem)
+              link.getTargetFileSystem();
+          try {
+            FileStatus status = linkedFs.getMyFs().getFileStatus(linkedPath);
+            result[i++] = new FileStatus(status.getLen(), false,
+              status.getReplication(), status.getBlockSize(),
+              status.getModificationTime(), status.getAccessTime(),
+              status.getPermission(), status.getOwner(), status.getGroup(),
+              link.getTargetLink(),
+              new Path(inode.fullPath).makeQualified(
+                  myUri, null));
+          } catch (FileNotFoundException ex) {
+            result[i++] = new FileStatus(0, false, 0, 0,
+              creationTime, creationTime, PERMISSION_555,
+              ugi.getShortUserName(), ugi.getPrimaryGroupName(),
+              link.getTargetLink(),
+              new Path(inode.fullPath).makeQualified(
+                  myUri, null));
```

Comment:
two comments:
1. seems to be some duplicate code, the "else" branch is pretty much the same, can we refactor here?
2. Not sure if this is the best way when dealing with FileNotFoundException. If I understand this correctly, it is possible that some mounts does not have this path, so it can hit FileNotFoundException? 

If this is the case, I wonder if it makes more sense to just skip this mount, by not adding a FileStatus for mount at all. So that clients do not get confused by an actually non-existing FileStatus, among other existing ones. But one issue here would be that result array is strictly the size of the # of mounts. Creating result as a list, append, and then return as a array may resolve this.

Reply:
For #2: The problem here I think we may not be able reset mount automatically, so until some one checks file exists or do op on target, we will not know whether the file exists or not. This will continue until user updates mount points accordingly.
This can be possible when some one deletes the target directory directly but not updated the mount tables accordingly. Please check one of my comment on behavior of ls in MAC. Also we have other issue: [ isDir is inconsistent](https://issues.apache.org/jira/browse/HDFS-15370?focusedCommentId=17116927&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17116927).

Reply:
Sorry if I am missing anything. The change with catching the `FileNotFoundException` is good ? or I need to make a change.

Reply:
Sorry if I am missing anything. The change with catching the FileNotFoundException is good ? or I need to change something. @umamaheswararao

Reply:
IMO, FileNotFoundException is fine. Please see my comment below https://github.com/apache/hadoop/pull/2019/files#r430611396
@chliang71 , What do you say?